File: java_projects/glue_code/commons-csv/src/main/java/org/apache/commons/csv/Lexer.java
package org.apache.commons.csv;

import static org.apache.commons.csv.Constants.CR;
import static org.apache.commons.csv.Constants.LF;

import java.io.Closeable;
import java.io.IOException;
import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;

final class Lexer implements Closeable {
  private static final char DISABLED = '\ufffe';
  private static final String LF_STRING = Character.toString(LF);
  private static final String CR_STRING = Character.toString(CR);
  private static Value clz = ContextInitializer.getPythonClass("/Lexer.py", "Lexer");
  private Value obj;

  public Lexer(Value obj) {
    this.obj = obj;
  }

  public Value getPythonObject() {
    return obj;
  }

  public void close() throws IOException {
    try {
      //
      // reader.close();
      //

      obj.invokeMember("close");
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "Lexer.close");
    }
  }

  private Token parseSimpleToken(final Token token, int ch) throws IOException {
    try {
      //
      // while (true) {
      // if (readEndOfLine(ch)) {
      // token.type = EORECORD;
      // break;
      // }
      // if (isEndOfFile(ch)) {
      // token.type = EOF;
      // token.isReady = true; // There is data at EOF
      // break;
      // }
      // if (isDelimiter(ch)) {
      // token.type = TOKEN;
      // break;
      // }
      // if (isEscape(ch)) {
      // if (isEscapeDelimiter()) {
      // token.content.append(delimiter);
      // } else {
      // final int unescaped = readEscape();
      // if (unescaped == END_OF_STREAM) { // unexpected char after escape
      // token.content.append((char) ch).append((char) reader.getLastChar());
      // } else {
      // token.content.append((char) unescaped);
      // }
      // }
      // } else {
      // token.content.append((char) ch);
      // }
      // ch = reader.read0(); // continue
      // }
      //
      // if (ignoreSurroundingSpaces) {
      // trimTrailingSpaces(token.content);
      // }
      //
      // return token;
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("parseSimpleToken", token, ch).as(Token.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "Lexer.parseSimpleToken");
    }
  }

  private Token parseEncapsulatedToken(final Token token) throws IOException {
    try {
      //
      // token.isQuoted = true;
      // final long startLineNumber = getCurrentLineNumber();
      // int c;
      // while (true) {
      // c = reader.read0();
      //
      // if (isEscape(c)) {
      // if (isEscapeDelimiter()) {
      // token.content.append(delimiter);
      // } else {
      // final int unescaped = readEscape();
      // if (unescaped == END_OF_STREAM) { // unexpected char after escape
      // token.content.append((char) c).append((char) reader.getLastChar());
      // } else {
      // token.content.append((char) unescaped);
      // }
      // }
      // } else if (isQuoteChar(c)) {
      // if (isQuoteChar(reader.lookAhead0())) {
      // c = reader.read0();
      // token.content.append((char) c);
      // } else {
      // while (true) {
      // c = reader.read0();
      // if (isDelimiter(c)) {
      // token.type = TOKEN;
      // return token;
      // }
      // if (isEndOfFile(c)) {
      // token.type = EOF;
      // token.isReady = true; // There is data at EOF
      // return token;
      // }
      // if (readEndOfLine(c)) {
      // token.type = EORECORD;
      // return token;
      // }
      // if (!Character.isWhitespace((char) c)) {
      // throw new IOException(
      // "(line "
      // + getCurrentLineNumber()
      // + ") invalid char between encapsulated token and"
      // + " delimiter");
      // }
      // }
      // }
      // } else if (isEndOfFile(c)) {
      // throw new IOException(
      // "(startline "
      // + startLineNumber
      // + ") EOF reached before encapsulated token finished");
      // } else {
      // token.content.append((char) c);
      // }
      // }
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("parseEncapsulatedToken", token).as(Token.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "Lexer.parseEncapsulatedToken");
    }
  }

  private char mapNullToDisabled(final Character c) {
    //
    // return c == null ? DISABLED : c.charValue();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("mapNullToDisabled", c).as(char.class);
  }

  private boolean isMetaChar(final int ch) {
    //
    // return ch == escape || ch == quoteChar || ch == commentStart;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("isMetaChar", ch).as(boolean.class);
  }

  void trimTrailingSpaces(final StringBuilder buffer) {
    //
    // int length = buffer.length();
    // while (length > 0 && Character.isWhitespace(buffer.charAt(length - 1))) {
    // length = length - 1;
    // }
    // if (length != buffer.length()) {
    // buffer.setLength(length);
    // }
    //

    obj.invokeMember("trimTrailingSpaces", buffer);
  }

  int readEscape() throws IOException {
    try {
      //
      // final int ch = reader.read0();
      // switch (ch) {
      // case 'r':
      // return CR;
      // case 'n':
      // return LF;
      // case 't':
      // return TAB;
      // case 'b':
      // return BACKSPACE;
      // case 'f':
      // return FF;
      // case CR:
      // case LF:
      // case FF: // TODO is this correct?
      // case TAB: // TODO is this correct? Do tabs need to be escaped?
      // case BACKSPACE: // TODO is this correct?
      // return ch;
      // case END_OF_STREAM:
      // throw new IOException("EOF whilst processing escape sequence");
      // default:
      // if (isMetaChar(ch)) {
      // return ch;
      // }
      // return END_OF_STREAM;
      // }
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("readEscape").as(int.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "Lexer.readEscape");
    }
  }

  boolean readEndOfLine(int ch) throws IOException {
    try {
      //
      // if (ch == CR && reader.lookAhead0() == LF) {
      // ch = reader.read0();
      // if (firstEol == null) {
      // this.firstEol = Constants.CRLF;
      // }
      // }
      // if (firstEol == null) {
      // if (ch == LF) {
      // this.firstEol = LF_STRING;
      // } else if (ch == CR) {
      // this.firstEol = CR_STRING;
      // }
      // }
      //
      // return ch == LF || ch == CR;
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("readEndOfLine", ch).as(boolean.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "Lexer.readEndOfLine");
    }
  }

  Token nextToken(final Token token) throws IOException {
    try {
      //
      //
      // int lastChar = reader.getLastChar();
      //
      // int c = reader.read0();
      // /*
      // * Note: The following call will swallow LF if c == CR. But we don't need to know if the
      // last char was CR or LF
      // * - they are equivalent here.
      // */
      // boolean eol = readEndOfLine(c);
      //
      // if (ignoreEmptyLines) {
      // while (eol && isStartOfLine(lastChar)) {
      // lastChar = c;
      // c = reader.read0();
      // eol = readEndOfLine(c);
      // if (isEndOfFile(c)) {
      // token.type = EOF;
      // return token;
      // }
      // }
      // }
      //
      // if (isEndOfFile(lastChar) || !isLastTokenDelimiter && isEndOfFile(c)) {
      // token.type = EOF;
      // return token;
      // }
      //
      // if (isStartOfLine(lastChar) && isCommentStart(c)) {
      // final String line = reader.readLine();
      // if (line == null) {
      // token.type = EOF;
      // return token;
      // }
      // final String comment = line.trim();
      // token.content.append(comment);
      // token.type = COMMENT;
      // return token;
      // }
      //
      // while (token.type == INVALID) {
      // if (ignoreSurroundingSpaces) {
      // while (Character.isWhitespace((char) c) && !isDelimiter(c) && !eol) {
      // c = reader.read0();
      // eol = readEndOfLine(c);
      // }
      // }
      //
      // if (isDelimiter(c)) {
      // token.type = TOKEN;
      // } else if (eol) {
      // token.type = EORECORD;
      // } else if (isQuoteChar(c)) {
      // parseEncapsulatedToken(token);
      // } else if (isEndOfFile(c)) {
      // token.type = EOF;
      // token.isReady = true; // there is data at EOF
      // } else {
      // parseSimpleToken(token, c);
      // }
      // }
      // return token;
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("nextToken", token).as(Token.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "Lexer.nextToken");
    }
  }

  boolean isStartOfLine(final int ch) {
    //
    // return ch == LF || ch == CR || ch == UNDEFINED;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("isStartOfLine", ch).as(boolean.class);
  }

  boolean isQuoteChar(final int ch) {
    //
    // return ch == quoteChar;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("isQuoteChar", ch).as(boolean.class);
  }

  boolean isEscapeDelimiter() throws IOException {
    try {
      //
      // reader.lookAhead1(escapeDelimiterBuf);
      // if (escapeDelimiterBuf[0] != delimiter[0]) {
      // return false;
      // }
      // for (int i = 1; i < delimiter.length; i++) {
      // if (escapeDelimiterBuf[2 * i] != delimiter[i]
      // || escapeDelimiterBuf[2 * i - 1] != escape) {
      // return false;
      // }
      // }
      // final int count = reader.read1(escapeDelimiterBuf, 0, escapeDelimiterBuf.length);
      // return count != END_OF_STREAM;
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("isEscapeDelimiter").as(boolean.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "Lexer.isEscapeDelimiter");
    }
  }

  boolean isEscape(final int ch) {
    //
    // return ch == escape;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("isEscape", ch).as(boolean.class);
  }

  boolean isEndOfFile(final int ch) {
    //
    // return ch == END_OF_STREAM;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("isEndOfFile", ch).as(boolean.class);
  }

  boolean isDelimiter(final int ch) throws IOException {
    try {
      //
      // isLastTokenDelimiter = false;
      // if (ch != delimiter[0]) {
      // return false;
      // }
      // if (delimiter.length == 1) {
      // isLastTokenDelimiter = true;
      // return true;
      // }
      // reader.lookAhead1(delimiterBuf);
      // for (int i = 0; i < delimiterBuf.length; i++) {
      // if (delimiterBuf[i] != delimiter[i + 1]) {
      // return false;
      // }
      // }
      // final int count = reader.read1(delimiterBuf, 0, delimiterBuf.length);
      // isLastTokenDelimiter = count != END_OF_STREAM;
      // return isLastTokenDelimiter;
      //

      // TODO: Check the type mapping below!
      return obj.invokeMember("isDelimiter", ch).as(boolean.class);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "Lexer.isDelimiter");
    }
  }

  boolean isCommentStart(final int ch) {
    //
    // return ch == commentStart;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("isCommentStart", ch).as(boolean.class);
  }

  boolean isClosed() {
    //
    // return reader.isClosed();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("isClosed").as(boolean.class);
  }

  String getFirstEol() {
    //
    // return firstEol;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getFirstEol").as(String.class);
  }

  long getCurrentLineNumber() {
    //
    // return reader.getCurrentLineNumber();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getCurrentLineNumber").as(long.class);
  }

  long getCharacterPosition() {
    //
    // return reader.getPosition();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getCharacterPosition").as(long.class);
  }

  Lexer(final CSVFormat format, final ExtendedBufferedReader reader) {
    //
    // this.reader = reader;
    // this.delimiter = format.getDelimiterString().toCharArray();
    // this.escape = mapNullToDisabled(format.getEscapeCharacter());
    // this.quoteChar = mapNullToDisabled(format.getQuoteCharacter());
    // this.commentStart = mapNullToDisabled(format.getCommentMarker());
    // this.ignoreSurroundingSpaces = format.getIgnoreSurroundingSpaces();
    // this.ignoreEmptyLines = format.getIgnoreEmptyLines();
    // this.delimiterBuf = new char[delimiter.length - 1];
    // this.escapeDelimiterBuf = new char[2 * delimiter.length - 1];
    //

    this.obj = clz.invokeMember("__init__", format, reader);
  }
}

File: data/recomposed_projects/deepseek-coder-33b-instruct/body/commons-csv/src/main/org/apache/commons/csv/Lexer.py
from __future__ import annotations
import re
import io
import numbers
import typing
from typing import *
import os
from src.main.org.apache.commons.csv.CSVFormat import *
from src.main.org.apache.commons.csv.Constants import *
from src.main.org.apache.commons.csv.ExtendedBufferedReader import *
from src.main.org.apache.commons.csv.Token import *


class Lexer:

    __isLastTokenDelimiter: bool = False

    __firstEol: str = ""

    __reader: ExtendedBufferedReader = None

    __ignoreEmptyLines: bool = False

    __ignoreSurroundingSpaces: bool = False

    __commentStart: str = ""

    __quoteChar: str = ""

    __escape: str = ""

    __escapeDelimiterBuf: typing.List[str] = None

    __delimiterBuf: typing.List[str] = None

    __delimiter: typing.List[str] = None

    __DISABLED: str = "\ufffe"
    __LF_STRING: str = "\n"
    __CR_STRING: str = "\r"

    def close(self) -> None:
        self.__reader.close()

    def __parseSimpleToken(self, token: Token, ch: int) -> Token:

        while True:
            if self.readEndOfLine(ch):
                token.type = EORECORD
                break
            if self.isEndOfFile(ch):
                token.type = EOF
                token.isReady = True  # There is data at EOF
                break
            if self.isDelimiter(ch):
                token.type = TOKEN
                break
            if self.isEscape(ch):
                if self.isEscapeDelimiter():
                    token.content.append(self.__delimiter)
                else:
                    unescaped = self.readEscape()
                    if unescaped == END_OF_STREAM:  # unexpected char after escape
                        token.content.append(chr(ch) + chr(self.__reader.getLastChar()))
                    else:
                        token.content.append(chr(unescaped))
            else:
                token.content.append(chr(ch))
            ch = self.__reader.read0()  # continue

        if self.__ignoreSurroundingSpaces:
            self.trimTrailingSpaces(token.content)

        return token

    def __parseEncapsulatedToken(self, token: Token) -> Token:

        token.isQuoted = True
        startLineNumber = self.getCurrentLineNumber()
        c = None

        while True:
            c = self.__reader.read0()

            if self.isEscape(c):
                if self.isEscapeDelimiter():
                    token.content += self.__delimiter
                else:
                    unescaped = self.readEscape()
                    if unescaped == END_OF_STREAM:
                        token.content += chr(c) + chr(self.__reader.getLastChar())
                    else:
                        token.content += chr(unescaped)
            elif self.isQuoteChar(c):
                if self.isQuoteChar(self.__reader.lookAhead0()):
                    c = self.__reader.read0()
                    token.content += chr(c)
                else:
                    while True:
                        c = self.__reader.read0()
                        if self.isDelimiter(c):
                            token.type = TOKEN
                            return token
                        if self.isEndOfFile(c):
                            token.type = EOF
                            token.isReady = True
                            return token
                        if self.readEndOfLine(c):
                            token.type = EORECORD
                            return token
                        if not chr(c).isspace():
                            raise IOError(
                                "(line "
                                + str(self.getCurrentLineNumber())
                                + ") invalid char between encapsulated token and"
                                + " delimiter"
                            )
            elif self.isEndOfFile(c):
                raise IOError(
                    "(startline "
                    + str(startLineNumber)
                    + ") EOF reached before encapsulated token finished"
                )
            else:
                token.content += chr(c)

    def __mapNullToDisabled(self, c: str) -> str:
        return self.__DISABLED if c is None else c

    def __isMetaChar(self, ch: int) -> bool:
        return (
            ch == ord(self.__escape)
            or ch == ord(self.__quoteChar)
            or ch == ord(self.__commentStart)
        )

    def trimTrailingSpaces(self, buffer: str) -> None:

        length = len(buffer)
        while length > 0 and buffer[length - 1].isspace():
            length = length - 1
        if length != len(buffer):
            buffer = buffer[:length]

    def readEscape(self) -> int:

        ch = self.__reader.read0()

        if ch == ord("r"):
            return ord(self.__CR_STRING)
        elif ch == ord("n"):
            return ord(self.__LF_STRING)
        elif ch == ord("t"):
            return ord(TAB)
        elif ch == ord("b"):
            return ord(BACKSPACE)
        elif ch == ord("f"):
            return ord(FF)
        elif (
            ch == ord(CR)
            or ch == ord(LF)
            or ch == ord(FF)
            or ch == ord(TAB)
            or ch == ord(BACKSPACE)
        ):
            return ch
        elif ch == END_OF_STREAM:
            raise IOError("EOF whilst processing escape sequence")
        else:
            if self.__isMetaChar(ch):
                return ch
            else:
                return END_OF_STREAM

    def readEndOfLine(self, ch: int) -> bool:

        if ch == CR and self.__reader.lookAhead0() == LF:
            ch = self.__reader.read0()
            if self.__firstEol is None:
                self.__firstEol = CRLF

        if self.__firstEol is None:
            if ch == LF:
                self.__firstEol = self.__LF_STRING
            elif ch == CR:
                self.__firstEol = self.__CR_STRING

        return ch == LF or ch == CR

    def nextToken(self, token: Token) -> Token:

        lastChar = self.__reader.getLastChar()

        c = self.__reader.read0()
        eol = self.readEndOfLine(c)

        if self.__ignoreEmptyLines:
            while eol and self.isStartOfLine(lastChar):
                lastChar = c
                c = self.__reader.read0()
                eol = self.readEndOfLine(c)
                if self.isEndOfFile(c):
                    token.type = EOF
                    return token

        if (
            self.isEndOfFile(lastChar)
            or not self.__isLastTokenDelimiter
            and self.isEndOfFile(c)
        ):
            token.type = EOF
            return token

        if self.isStartOfLine(lastChar) and self.isCommentStart(c):
            line = self.__reader.readLine()
            if line is None:
                token.type = EOF
                return token
            comment = line.strip()
            token.content.append(comment)
            token.type = COMMENT
            return token

        while token.type == INVALID:
            if self.__ignoreSurroundingSpaces:
                while (
                    Character.isWhitespace(chr(c))
                    and not self.isDelimiter(c)
                    and not eol
                ):
                    c = self.__reader.read0()
                    eol = self.readEndOfLine(c)

            if self.isDelimiter(c):
                token.type = TOKEN
            elif eol:
                token.type = EORECORD
            elif self.isQuoteChar(c):
                self.__parseEncapsulatedToken(token)
            elif self.isEndOfFile(c):
                token.type = EOF
                token.isReady = True  # there is data at EOF
            else:
                self.__parseSimpleToken(token, c)

        return token

    def isStartOfLine(self, ch: int) -> bool:
        return (
            ch == ord(self.__LF_STRING)
            or ch == ord(self.__CR_STRING)
            or ch == UNDEFINED
        )

    def isQuoteChar(self, ch: int) -> bool:
        return ch == ord(self.__quoteChar)

    def isEscapeDelimiter(self) -> bool:

        self.__reader.lookAhead1(self.__escapeDelimiterBuf)

        if self.__escapeDelimiterBuf[0] != self.__delimiter[0]:
            return False

        for i in range(1, len(self.__delimiter)):
            if (
                self.__escapeDelimiterBuf[2 * i] != self.__delimiter[i]
                or self.__escapeDelimiterBuf[2 * i - 1] != self.__escape
            ):
                return False

        count = self.__reader.read1(
            self.__escapeDelimiterBuf, 0, len(self.__escapeDelimiterBuf)
        )

        return count != END_OF_STREAM

    def isEscape(self, ch: int) -> bool:
        return ch == ord(self.__escape)

    def isEndOfFile(self, ch: int) -> bool:
        return ch == END_OF_STREAM

    def isDelimiter(self, ch: int) -> bool:

        self.__isLastTokenDelimiter = False

        if ch != self.__delimiter[0]:
            return False

        if len(self.__delimiter) == 1:
            self.__isLastTokenDelimiter = True
            return True

        self.__reader.lookAhead1(self.__delimiterBuf)

        for i in range(len(self.__delimiterBuf)):
            if self.__delimiterBuf[i] != self.__delimiter[i + 1]:
                return False

        count = self.__reader.read1(self.__delimiterBuf, 0, len(self.__delimiterBuf))

        self.__isLastTokenDelimiter = count != END_OF_STREAM

        return self.__isLastTokenDelimiter

    def isCommentStart(self, ch: int) -> bool:
        return ch == ord(self.__commentStart)

    def isClosed(self) -> bool:
        return self.__reader.isClosed()

    def getFirstEol(self) -> str:
        return self.__firstEol

    def getCurrentLineNumber(self) -> int:

        return self.__reader.getCurrentLineNumber()

    def getCharacterPosition(self) -> int:
        return self.__reader.getPosition()

    def __init__(self, format: CSVFormat, reader: ExtendedBufferedReader) -> None:
        self.__reader = reader
        self.__delimiter = list(format.getDelimiterString())
        self.__escape = self.__mapNullToDisabled(format.getEscapeCharacter())
        self.__quoteChar = self.__mapNullToDisabled(format.getQuoteCharacter())
        self.__commentStart = self.__mapNullToDisabled(format.getCommentMarker())
        self.__ignoreSurroundingSpaces = format.getIgnoreSurroundingSpaces()
        self.__ignoreEmptyLines = format.getIgnoreEmptyLines()
        self.__delimiterBuf = [None] * (len(self.__delimiter) - 1)
        self.__escapeDelimiterBuf = [None] * (2 * len(self.__delimiter) - 1)
